File: src/medium/Binary_Tree_Level_Order_Traversal.java
package medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Binary_Tree_Level_Order_Traversal {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    class Solution {
        Map<Integer, List<Integer>> map = new HashMap<>();

        public List<List<Integer>> levelOrder(TreeNode root) {
            List<List<Integer>> ret = new ArrayList<>();
            levelOrder(root, 0);
            for(Map.Entry<Integer, List<Integer>> entry : map.entrySet()) {
                ret.add(entry.getValue());
            }
            return ret;
        }

        void levelOrder(TreeNode root, int level) {
            if (root == null) return;

            List<Integer> list = map.getOrDefault(level, new ArrayList<>());
            list.add(root.val);
            map.put(level, list);
            levelOrder(root.left, level + 1);
            levelOrder(root.right, level + 1);
        }
    }
}

File: src/medium/binary_tree_level_order_traversal.cpp
https://leetcode.com/problems/binary-tree-level-order-traversal

//using Queue

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        queue<TreeNode*> q;
        vector<vector<int>> v;
        vector<int> x;
        
        if(!root)
            return v;
        
        q.push(root);
        
        while(!q.empty())
        {
            int c = q.size();
            
            while(c--)
            {
                TreeNode* y = q.front();
                x.push_back(y->val);
                
                if(y->left) 
                    q.push(y->left);
                
                if(y->right) 
                    q.push(y->right);
                
                q.pop();
            }
             
            v.push_back(x);
            x.clear();
        }
        
        
        return v;
    }
};
